### ISSUE TITLE
keystore password hardcoded, no overload from config file

### ISSUE THREAD

**Issue by patrick92100** [gocd/gocd#1769]
security breach...

**Comment by arvindsv** [comment]
"Breach". What do you mean?

Of course, if you have access to the file, on the Go Server itself, you can see the certificates. But, that's hardly a "breach". This information is also quite widely published. For instance, [here](https://www.go.cd/2014/06/05/using-go-cd-with-custom-certificates.html) and on many instances in the mailing list.

I guess you mean an override from the config. In my view, that's insecure too and screams of security theatre. If you have someone who can access the keystore physically, there's nothing stopping them from accessing the config file physically and then opening the keystore.

As long as the Go Server needs to open it, that information will need to be made available to it and will be available where it exists. If the physical security of that file is suspect, then this will never be secure. In such cases, it is recommended to not try and have security through obscurity, but to be open about the fact that this information is insecure.

In summary, the vector of this "attack" is not going to be mitigated by putting that password in a config file.

So, based on this understanding of mine, I'm going to close this. Happy to reopen it if needed. If anyone does want to contribute a change, so that it gets this value from a config file, then it will probably get accepted. But, it'll need to take care of existing users.

**Comment by patrick92100** [comment]
Strange vision of security
why there is an entry in the jetty config file and dont use it? 
why should I rebuild the server/agent to change a password? It was like this 20 years ago...
the issue, it is not where is stored the password, the issue is the password is the same on all GO Server/Agent instances used by every one, untill they re compile to change the password...

**Comment by lenucksi** [comment]
An additional question here is: Which keystore? There is at least a server and a client keystore, both of which passwords are hardcoded. 
The client uses it to store it client certificates with which is authenticates against the server and and the server stores at least its certicate that is presented to the clients. 

While mutual SSL authentication between agent and server is a good idea, changing the certificate in the server is quite a bit of a annoyance, especially since not every valid SSL certificate is accepted, the exchange process is tedious and requires a restart of the server.
Also it mandates the opening of ports other than 443 as this authentication using client certificates can not reasonably by proxied through a standard webserver, some of it being due to enforcement in code.

As @arvindsv says: I you are able to retrieve the keystore on either Server or Agent you have either found a real bug in Go that allowed file system access or worse or have shell access to the machine in question. In any of those cases having the password to the keystore is the smaller problem than the one that got you the access. Especially since you would probably have the password from the config file then anyway.

I guess a more interesting issue might be what arises from the lack of documentation of the authentication process between agent and server, the tediousness of exchanging the certificates and inability to proxy Go entirely through a standard proxy...

**Comment by patrick92100** [comment]
both keystore, Agent and Server
make a search on
serverKeystorepa55w0rd
agent5s0repa55w0rd
I put my own certificate on the server and agent side, which is the minimun what everyone should do to set up security...

**Comment by arvindsv** [comment]
> Strange vision of security

It's not a "vision". It's security by obscurity to do what you say. If you store a password hard-coded or hard-coded next to where you use it, as @lenucksi said, the attack vector is the same. If you have access to that file, you have access to the password.

All I am saying is, it will give a false sense of security. It will not be any more secure. As I also said, I'm sure many people will "feel" more secure about it and so, a contribution towards this will probably be accepted, but security doesn't work like that.

**Comment by patrick92100** [comment]
it always a bad idea to implement your own security and do not use correctly the one proposed by the framework, if you think you do better than the jetty framework, go ahead...

**Comment by patrick92100** [comment]
so for you, if I get your point, when you have to change your login password on your PC, you should recompile the OS? nobody do that...

**Comment by arvindsv** [comment]
This is getting towards trolling and I'll probably not reply for too much longer. But, this is [your argument](https://en.wikipedia.org/wiki/Straw_man). I never really mentioned logging in anywhere. I'd suggest that there's a difference between a user having a password in their head and a server which needs to come up unattended and have a password available to it.

**Comment by arvindsv** [comment]
And no, you don't get my point.

**Comment by grahamc** [comment]
@patrick92100: As far as I understand it, if the password is stored in a configuration file which GoCD can read (which it must, in order to be able to get the password) and the keystore etc. are also readable by GoCD, then there is no practical difference between having a hardcoded or not hardcoded password: an attacker with access to the keystore can also access the config file. Do you believe otherwise?

**Comment by patrick92100** [comment]
@grahamc: No, what I want is to be able to easily change the keystore password at Server and Agent side.
my point is:
to set up the security on GO Server and Agent, 
1- You should set up your own certificate chain, signed with a valid CA,
2- to set up your new certificate chain, you have to set it in the Server and Agent Keystore, to do this you should know the keystore password(I first look in the jetty config file and try to use this password=> not OK, I had to look into the src code to find the pwd...).
3- if you want your own keystore password(which is best), today in GO, you should re-compile the server and the agent (which is not the best practice...) 

If you estimate that there is no issue, please make a doc to explain how it works.

Thanks for your time

**Comment by patrick92100** [comment]
and, to me it is a security breach, because if it is to complicated to change the password of the keystore, nobody will change the Agent and the Server keystore password, and the passwords used will be the ones exposed in clear on the public repository. 
I agreed it is not a big security breach, but the security is weakened

**Comment by grahamc** [comment]
I understand what you want to do with the certificates. A quick google search turned up https://www.go.cd/2014/06/05/using-go-cd-with-custom-certificates.html which documents the process. If these docs aren't in the standard docs, they might should be copied over.

On the part of a configurable keystore password, can you find any best practices documentation about the Keystore password when deployed to servers?

I went ahead and asked on Freenode in `##java` about using keystore in production and how to deal with the passwords, and this is what they said:

```
gchristensen | Hi, a project I use uses Keystore with a built-in, default, standard
             | password across all installations. It is software typically installed on
             | headless servers. The reasoning for not making the password configurable is
             | because if the server can read the password and the keystore, any
             | compromise on the server which can get one can get both. 1. does this make
             | sense? 2. are there any best practices about keystores
gchristensen | and password management in a headless server, or integration with HMSs or
             | something?
     selckin | makes sense to me
     selckin | if you tape the key next to the lock on your door, does it matter if you
             | lock it?
```

They then followed up with information that Java supports a hardware security module with PKCS 11. This would be an _extremely_ interesting addition to Go.CD and I'm sure the team would work with you (or anyone else) to implement it.

You can find some more information here:
 - https://docs.oracle.com/javase/7/docs/technotes/guides/security/p11guide.html
 - http://www.pixelstech.net/article/1420699130-Different-types-of-keystore-in-Java----PKCS11
 - https://en.wikipedia.org/wiki/PKCS_11
 - logs from ##java: https://javabot.evanchooly.com/logs/%23%23java/2016-01-08 09:00

**Comment by patrick92100** [comment]
Yes, I saw the page afterwards ;)
The jetty framework embed all the process, there is not much to do,
http://www.eclipse.org/jetty/documentation/current/configuring-ssl.html 
look at 'Configuring SslContextFactory'
normally you just have to set correctly the following file
http://git.eclipse.org/c/jetty/org.eclipse.jetty.project.git/plain/jetty-server/src/main/config/etc/jetty-ssl-context.xml
the password are ofuscated and can be generated with org.mortbay.jetty.security.Password class
http://stackoverflow.com/questions/4690970/keystore-password-for-jetty-7-0-ssl-configuration

**Comment by hprasad068** [comment]
@patrick92100  have you noticed trust.jks on the agent config's when you configured your own cert? i'm not sure if that's needed or how it is getting created. we're on 17.2.0 version
